File: project-euler/python/012.py
# This passes the first 4 cases but timesout otherwise...?
# This implementation replicates what I have done in JS

#!/bin/python3
import sys
from functools import reduce

def triangular_n(n):
    return int(n*(n+1)/2)

def prime_factorisation(n):
    if n == 1:
        return { 1: 0 }
    prime_factors = {}
    while n%2 == 0:
        prime_factors[2] = prime_factors.get(2, 0) + 1
        n/=2
    for i in range(3, int(n**0.5)+1, 2):
        while (n%i == 0 and n != i):
            prime_factors[i] = prime_factors.get(i, 0) + 1
            n = n / i
    if n > 1:
        prime_factors[n] = prime_factors.get(n, 0) + 1
    return prime_factors

def highly_divisible_triangular_n(n):
    num_divisors = 0
    index = 0
    while num_divisors < (n+1):
        index+=1
        prime_factors = prime_factorisation(triangular_n(index))
        num_divisors = reduce(lambda x, y: x*(y+1), prime_factors.values(), 1)
        # print(index, triangular_n(index), prime_factors, list(prime_factors.values()), num_divisors)
    return triangular_n(index)

num_inputs = int(input().strip())
for _ in range(num_inputs):
    n = int(input().strip())
    print(highly_divisible_triangular_n(n))

File: project-euler/js/012.js
const { parse } = require('path');
const readline = require('readline')

const rl = readline.createInterface({input: process.stdin, output: process.stdout});
rl.prompt();

let input = '' 

rl.on('line', (inputLine) => {
    input+=(inputLine+'\n');
})

rl.on('close', () => {
    main(input);
    process.exit(0);
})

function triangularN(n) {
    return n*(n+1)/2
}

// Note you are doing this by brute force...
function numDivisors(n) {
    let arr = [1, n] //this handles the special case of n=1
    for (let i = 2; i <= Math.ceil(Math.sqrt(n)) + 1; i++) {
        if (n%i == 0) {
            arr.push(i, n/i) //there are a lot of duplicates?
        }
    }
    console.log([...new Set(arr)])
    return [...new Set(arr)].length
}

function getObjectValue(object, value) {
    return !!object[value] ? object[value] : 0
}

// Note: If the prime factors of a number are a^n * b^m * c^q, then the number of divisors is (n+1)(m+1)(q+1)
// https://www2.math.upenn.edu/~deturck/m170/wk2/numdivisors.html
function primeFactorisation(N) {
    let primeFactors = {} 
    while (N%2 === 0) {
        primeFactors[2] = getObjectValue(primeFactors, 2) + 1
        N = N / 2
    }
    for (let i = 3; i <= Math.ceil(Math.sqrt(N)) + 1; i+=2) { //note that keeping Math.sqrt(temp) actually makes a difference
        while (N%i === 0 && N !== i) {
            primeFactors[i] = getObjectValue(primeFactors, i) + 1
            N = N / i
        }        
    }
    if (N > 1) {
        primeFactors[N] = getObjectValue(primeFactors, N) + 1
    }
    return primeFactors;
}

function highlyDivTriangularN(N) {
    let numDivs = 0
    let index = 0
    while (numDivs < (N + 1)) {
        index++
        // numDivs = numDivisors(triangularN(index))
        primeFactors = primeFactorisation(triangularN(index))
        numDivs = Object.values(primeFactors).reduce((prod, elem) => prod*(elem+1), 1)
    }
    return triangularN(index)
}

function main(input) {
    const inputArray = input
        .split('\n')
        .filter(elem => elem !== '')
        .map(curr => parseInt(curr) )
    const numInputs = inputArray[0];

    for(var i = 1; i < numInputs+1; i++) {
        console.log(highlyDivTriangularN(inputArray[i]))
    }
}
